jax/experimental/jax2tf/README.md

```diff
@@ -595,16 +595,50 @@ The solution is to avoid `np.array`, `float`, or JAX arrays in operations whose
 results are used as shapes, e.g., instead of `np.arange(n) * x.shape[0]` write
 `[i * x.shape[0] for i in range(n)]`.
 
+### Dimension variables must be solvable from the input shapes
+
+JAX will generate code to derive the values of the dimension variables
+from the input shapes. This works only if the symbolic dimensions in the input shapes are linear.
+For example, the following `polymorphic_shapes` will result in errors:
+
+```python
+polymorphic_shapes = ["a * a"]  # Not a linear polynomial
+polymorphic_shapes = ["a + b"]  # Too few equations to derive both `a` and `b`
+```
+
+The error message for the last specification above would be:
+
+```
+Cannot solve for values of dimension variables {'a', 'b'}. "
+We can only solve linear uni-variate constraints. "
+Unprocessed specifications: 'a + b' for dimension size args[0].shape[0], . "
+Please see https://github.com/google/jax/blob/main/jax/experimental/jax2tf/README.md#dimension-variables-must-be-solvable-from-the-input-shapes for more details.
+```
+
+### Shape assertion errors
+
 JAX assumes that dimension variables range over strictly positive integers.
-These assumptions are now checked against the shapes of the actual arguments
+Starting with serialization version 7 these assumptions are
+checked against the shapes of the actual arguments
 when the lowered code is invoked.
 For example, given the `polymorphic_shapes="(b, b, 2*d)"`
 specification, we will generate code to check the following constraints when
 invoked with actual argument `arg`:
 
   * `arg.shape[0] >= 1`
   * `arg.shape[1] == arg.shape[0]`
-  * `arg.shape[2] % 2 == 0` and `arg.shape[2] // 2 >= 1`
+  * `arg.shape[2] % 2 == 0`
+  * `arg.shape[2] // 2 >= 1`
+
+An example error for the third constraint above, e.g., when invoked with
+shape `(3, 3, 5)`, would be:
+
+```
+Input shapes do not match the polymorphic shapes specification.
+Division had remainder 1 when computing the value of 'd'.
+Obtained dimension variables: 'b' = 3 from specification 'b' for dimension args[0].shape[0] (= 3)
+Please see https://github.com/google/jax/blob/main/jax/experimental/jax2tf/README.md#shape-assertion-errors for more details.
+```
```

Comment:
I think that you are right it would help. Just yesterday I was debugging one of these failures and I found helpful to see the polymorphic shape specification. I have added it, with one caveat. The API allows shortcuts, such as, "b, ...", or 
b, _", but the error message includes the fully elaborated specification, e.g., "b, 16".